rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,358 @@
+
+### Nonfunctional requirements
+
+1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
+1. A user can upload a single file up to 30GB in size.
+1. Shared links referring to more than one file will be rejected.
+1. The status of an upload must be tracked and queryable.
+1. Only .h5ad files are accepted.
+
+## Detailed Design | Architecture | Implementation
+
+![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
+Figure 1: The Upload Design
+
+### Cloud to Cloud Upload Flow
+
+Cloud to cloud upload requires the user to provide a shareable link for a file on their DropBox from which DP can upload
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 3.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 3:** Architecture of a user uploading from their cloud to our cloud.
+
+#### 1. Get Share Link
+
+The user must generate a sharable link from their DropBox. Sharable links should only point to a single file, not a
+directory of files.
+
+#### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
+check the status of the upload
+
+#### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by DropBox are often not direct
+links to the files and need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of the sanitization process. The backend that runs this should also be given minimal privileges to further protect from
+server-side request forgery. How to download from a shared link on [Dropbox](https://help.dropbox.com/files-folders/share/force-download).
+
+The backend will also create an entry in the upload table using the dataset_id as the key, and set the status to waiting.
+
+#### 4. Queue
+
+The queue stores the links that need to be processed. It buffers the downstream servers from download requests. This
+allows us to spin up servers as needed. It also gives us the ability to retry a download if one of the servers to fails
+upload the file.
+
+##### Why might the download fail
+
+- The downloading server ran out of storage.
+- The downloading server crashed.
+- DropBox experience and outage
+- The user has reached their daily bandwidth allowance from their DropBox. [more info](https://help.dropbox.com/files-folders/share/banned-links)
+
+#### 5. Start an Instance to Download File
+
+An EC2 instance, docker container, or AWS lambda can be used to complete this task. Once a compute resource is spawned,
```

Comment:
I suggest not using AWS lambda, just to avoid concerns with the time limit.

Reply:
I think using a container is the best solution. I don't have a lot of experience in using them, so I'll need to some help if I work on it.

Reply:
cc @ryanking Do you have a recommendation here?

Reply:
Yeah I recommend you talk to @jgadling :)

Jess, Mike and others are working on making it quite easy to deploy containers via ECS.

Reply:
Yep, we're supporting ECS going forward! There are a few different ways this workflow can be implemented on ECS depending on latency needs - let's talk about this when you're ready.